File: server/src/main/java/dev/ascape/game/world/TileMap.java
package dev.ascape.game.world;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Immutable grid parsed from a {@code shared/maps/*.map.txt} file. Outside the map counts as wall. */
public final class TileMap {

	private static final char OUTSIDE_CHAR = '#';

	private final String id;
	private final List<String> rows;
	private final TileType[][] types;
	private final TileType outsideType;
	private final int width;
	private final int height;

	private TileMap(String id, List<String> rows, TileType[][] types, TileType outsideType) {
		this.id = id;
		this.rows = List.copyOf(rows);
		this.types = types;
		this.outsideType = outsideType;
		this.height = rows.size();
		this.width = rows.getFirst().length();
	}

	public static TileMap parse(String id, String source, TileLegend legend) {
		List<String> rows = new ArrayList<>(Arrays.asList(source.replace("\r\n", "\n").split("\n", -1)));
		while (!rows.isEmpty() && rows.getLast().isBlank()) {
			rows.removeLast();
		}
		if (rows.isEmpty()) {
			throw new IllegalArgumentException("Map '" + id + "' is empty");
		}
		int width = rows.getFirst().length();
		TileType[][] types = new TileType[rows.size()][width];
		for (int y = 0; y < rows.size(); y++) {
			String row = rows.get(y);
			if (row.length() != width) {
				throw new IllegalArgumentException("Map '%s' row %d has width %d, expected %d"
						.formatted(id, y + 1, row.length(), width));
			}
			for (int x = 0; x < width; x++) {
				char mapChar = row.charAt(x);
				int column = x + 1;
				int line = y + 1;
				types[y][x] = legend.typeOf(mapChar)
						.orElseThrow(() -> new IllegalArgumentException("Map '%s' has unknown character '%s' at row %d, column %d"
								.formatted(id, mapChar, line, column)));
			}
		}
		TileType outside = legend.typeOf(OUTSIDE_CHAR)
				.orElseThrow(() -> new IllegalStateException("Tile legend lacks '" + OUTSIDE_CHAR + "'"));
		return new TileMap(id, rows, types, outside);
	}

	public String id() {
		return id;
	}

	/** Raw map lines, as sent to clients. */
	public List<String> rows() {
		return rows;
	}

	public int width() {
		return width;
	}

	public int height() {
		return height;
	}

	public boolean isInside(int x, int y) {
		return x >= 0 && y >= 0 && x < width && y < height;
	}

	public TileType typeAt(int x, int y) {
		return isInside(x, y) ? types[y][x] : outsideType;
	}

	public boolean isSolid(int x, int y) {
		return typeAt(x, y).solid();
	}

	public boolean isOpaque(int x, int y) {
		return typeAt(x, y).opaque();
	}

	public TileKind kindAt(int x, int y) {
		return typeAt(x, y).kind();
	}

	public List<GridPos> findAll(TileKind kind) {
		List<GridPos> positions = new ArrayList<>();
		for (int y = 0; y < height; y++) {
			for (int x = 0; x < width; x++) {
				if (types[y][x].kind() == kind) {
					positions.add(new GridPos(x, y));
				}
			}
		}
		return positions;
	}
}

File: client/src/game/tileMap.ts
import { TILE_DEFINITIONS, type TileDefinition, type TileKind } from './tiles'

export interface GridPosition {
  readonly x: number
  readonly y: number
}

export class TileMap {
  readonly width: number
  readonly height: number
  private readonly rows: readonly string[]

  private constructor(rows: readonly string[]) {
    this.rows = rows
    this.height = rows.length
    this.width = rows[0].length
  }

  /**
   * Parses a map file (`shared/maps/*.map.txt`): one row per line, every row the same width,
   * every character a key of TILE_DEFINITIONS. Trailing blank lines are ignored.
   */
  static parse(source: string): TileMap {
    const rows = source.replace(/\r\n/g, '\n').split('\n')
    while (rows.length > 0 && rows[rows.length - 1].trim() === '') rows.pop()
    if (rows.length === 0) throw new Error('Map is empty')

    const width = [...rows[0]].length
    rows.forEach((row, y) => {
      const chars = [...row]
      if (chars.length !== width) {
        throw new Error(`Map row ${y + 1} has width ${chars.length}, expected ${width}`)
      }
      chars.forEach((char, x) => {
        if (!(char in TILE_DEFINITIONS)) {
          throw new Error(`Unknown map character '${char}' at row ${y + 1}, column ${x + 1}`)
        }
      })
    })
    return new TileMap(rows)
  }

  isInside(x: number, y: number): boolean {
    return x >= 0 && y >= 0 && x < this.width && y < this.height
  }

  /** Map character at (x, y); outside the map counts as wall. */
  charAt(x: number, y: number): string {
    return this.isInside(x, y) ? this.rows[y][x] : '#'
  }

  tileAt(x: number, y: number): TileDefinition {
    return TILE_DEFINITIONS[this.charAt(x, y)]
  }

  isSolid(x: number, y: number): boolean {
    return this.tileAt(x, y).solid
  }

  isOpaque(x: number, y: number): boolean {
    return this.tileAt(x, y).opaque
  }

  findAll(kind: TileKind): GridPosition[] {
    const positions: GridPosition[] = []
    this.rows.forEach((row, y) => {
      for (let x = 0; x < row.length; x++) {
        if (TILE_DEFINITIONS[row[x]].kind === kind) positions.push({ x, y })
      }
    })
    return positions
  }
}
